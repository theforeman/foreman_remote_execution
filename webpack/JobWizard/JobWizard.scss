.job-wizard {
  .pf-c-wizard__main {
    overflow: visible;
    z-index: calc(
      var(--pf-c-wizard__footer--ZIndex) + 1
    ); // So the select box can be shown above the wizard footer
  }

  .pf-c-wizard__main-body {
    max-width: 500px;
  }
}
