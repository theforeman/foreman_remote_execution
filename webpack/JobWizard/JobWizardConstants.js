import { foremanUrl } from 'foremanReact/common/helpers';

export const JOB_TEMPLATES = 'JOB_TEMPLATES';
export const JOB_CATEGORIES = 'JOB_CATEGORIES';
export const templatesUrl = foremanUrl('/api/v2/job_templates');
