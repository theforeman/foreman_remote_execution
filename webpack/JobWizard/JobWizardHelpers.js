/* eslint-disable camelcase */
export const generateDefaultDescription = ({
  description_format,
  advancedInputs,
  inputs,
  name,
}) => {
  if (description_format) return description_format;
  const allInputs = [...advancedInputs, ...inputs];
  if (!allInputs.length) return name;
  const inputsString = allInputs
    .map(({ name: inputname }) => `${inputname}="%{${inputname}}"`)
    .join(' ');
  return `${name} with inputs ${inputsString}`;
};
