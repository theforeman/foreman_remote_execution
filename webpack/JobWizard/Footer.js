import React from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Button,
  WizardFooter,
  WizardContextConsumer,
  Tooltip,
  Spinner,
} from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import { WIZARD_TITLES } from './JobWizardConstants';
import { selectIsSubmitting } from './JobWizardSelectors';

export const Footer = ({ canSubmit, onSave }) => {
  const isSubmitting = useSelector(selectIsSubmitting);
  return (
    <WizardFooter>
      <WizardContextConsumer>
        {({ activeStep, onNext, onBack, onClose, goToStepByName }) => {
          const isValid =
            activeStep && activeStep.enableNext !== undefined
              ? activeStep.enableNext
              : true;

          return (
            <>
              <Button
                variant="primary"
                type="submit"
                onClick={onNext}
                isDisabled={!isValid || isSubmitting}
              >
                {activeStep.name === WIZARD_TITLES.review
                  ? __('Submit')
                  : __('Next')}
              </Button>
              <Button
                variant="secondary"
                onClick={onBack}
                isDisabled={
                  activeStep.name === WIZARD_TITLES.categoryAndTemplate
                }
              >
                {__('Back')}
              </Button>
              <Tooltip
                content={
                  <div>
                    {canSubmit
                      ? __('Start job')
                      : __('Fill all required fields in all the steps')}
                  </div>
                }
              >
                <Button
                  variant="tertiary"
                  onClick={onSave}
                  isAriaDisabled={!canSubmit}
                  isDisabled={isSubmitting}
                >
                  {__('Run on selected hosts')}
                </Button>
              </Tooltip>
              <Tooltip
                content={
                  <div>
                    {canSubmit
                      ? __('Skip to review step')
                      : __(
                          'Fill all required fields in all the steps to start the job'
                        )}
                  </div>
                }
              >
                <Button
                  variant="tertiary"
                  onClick={() => goToStepByName(WIZARD_TITLES.review)}
                  isAriaDisabled={!canSubmit}
                  isDisabled={isSubmitting}
                >
                  {__('Skip to review')}
                </Button>
              </Tooltip>
              <Button variant="link" onClick={onClose}>
                {__('Cancel')}
              </Button>
              {isSubmitting && (
                <div>
                  <Spinner size="lg" />
                </div>
              )}
            </>
          );
        }}
      </WizardContextConsumer>
    </WizardFooter>
  );
};

Footer.propTypes = {
  canSubmit: PropTypes.bool.isRequired,
  onSave: PropTypes.func.isRequired,
};
