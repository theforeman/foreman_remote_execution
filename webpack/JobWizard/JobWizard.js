import React, { useState } from 'react';
import { Wizard } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import history from 'foremanReact/history';
import CategoryAndTemplate from './steps/CategoryAndTemplate/';
import './JobWizard.scss';

export const JobWizard = () => {
  const [jobTemplate, setJobTemplate] = useState(null);
  const [category, setCategory] = useState('');
  const steps = [
    {
      name: __('Category and template'),
      component: (
        <CategoryAndTemplate
          jobTemplate={jobTemplate}
          setJobTemplate={setJobTemplate}
          category={category}
          setCategory={setCategory}
        />
      ),
    },
    {
      name: __('Target hosts'),
      component: <p>TargetHosts </p>,
      canJumpTo: !!jobTemplate,
    },
    {
      name: __('Advanced fields'),
      component: <p> AdvancedFields </p>,
      canJumpTo: !!jobTemplate,
    },
    {
      name: __('Schedule'),
      component: <p>Schedule</p>,
      canJumpTo: !!jobTemplate,
    },
    {
      name: __('Review details'),
      component: <p>ReviewDetails</p>,
      nextButtonText: 'Run',
      canJumpTo: !!jobTemplate,
    },
  ];
  const title = __('Run Job');
  return (
    <Wizard
      onClose={() => history.goBack()}
      navAriaLabel={`${title} steps`}
      steps={steps}
      height="70vh"
      className="job-wizard"
    />
  );
};
