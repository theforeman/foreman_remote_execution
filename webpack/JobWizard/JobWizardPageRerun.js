import React from 'react';
import PropTypes from 'prop-types';
import URI from 'urijs';
import {
  Alert,
  Title,
  Divider,
  Skeleton,
  Flex,
  FlexItem,
  Button,
} from '@patternfly/react-core';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import { useAPI } from 'foremanReact/common/hooks/API/APIHooks';
import PageLayout from 'foremanReact/routes/common/PageLayout/PageLayout';
import { STATUS } from 'foremanReact/constants';
import {
  useForemanOrganization,
  useForemanLocation,
} from 'foremanReact/Root/Context/ForemanContext';
import { JobWizard } from './JobWizard';
import { JOB_API_KEY } from './JobWizardConstants';

const JobWizardPageRerun = ({
  match: {
    params: { id },
  },
  location: { search },
}) => {
  const uri = new URI(search);
  const { failed_only: failedOnly } = uri.search(true);
  const { response, status } = useAPI(
    'get',
    `/ui_job_wizard/job_invocation?id=${id}${
      failedOnly ? '&failed_only=1' : ''
    }`,
    JOB_API_KEY
  );
  const title = __('Run job');
  const breadcrumbOptions = {
    breadcrumbItems: [
      { caption: __('Jobs'), url: `/job_invocations` },
      { caption: title },
    ],
  };

  const jobOrganization = response.job_organization;
  const jobLocation = response.job_location;
  const currentOrganization = useForemanOrganization();
  const currentLocation = useForemanLocation();

  return (
    <PageLayout
      header={title}
      breadcrumbOptions={breadcrumbOptions}
      searchable={false}
    >
      <React.Fragment>
        <React.Fragment>
          <Flex>
            <FlexItem>
              <Title headingLevel="h2" size="2xl">
                {title}
              </Title>
            </FlexItem>
            <FlexItem align={{ default: 'alignRight' }}>
              <Button
                variant="link"
                component="a"
                href={`/old/job_invocations/${id}/rerun${search}`}
              >
                {__('Use old form')}
              </Button>
            </FlexItem>
          </Flex>
          <Divider component="div" />
        </React.Fragment>
        {!status || status === STATUS.PENDING ? (
          <div style={{ height: '400px' }}>
            <Skeleton
              height="100%"
              screenreaderText="Loading large rectangle contents"
            />
          </div>
        ) : (
          <React.Fragment>
            {jobOrganization?.id !== currentOrganization?.id && (
              <Alert
                isInline
                className="job-wizard-alert"
                variant="warning"
                title={sprintf(
                  __(
                    "Current organization %s is different from job's organization %s. This job may run on different hosts than before.",
                    currentOrganization,
                    jobOrganization
                  )
                )}
              />
            )}
            {jobLocation?.id !== currentLocation?.id && (
              <Alert
                isInline
                className="job-wizard-alert"
                variant="warning"
                title={sprintf(
                  __(
                    "Current location %s is different from job's location %s. This job may run on different hosts than before.",
                    currentLocation,
                    jobLocation
                  )
                )}
              />
            )}
            <Divider component="div" />
            <JobWizard
              rerunData={{ ...response?.job, inputs: response?.inputs } || null}
            />
          </React.Fragment>
        )}
      </React.Fragment>
    </PageLayout>
  );
};
JobWizardPageRerun.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
  }).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
};
JobWizardPageRerun.defaultProps = {
  location: { search: '' },
};
export default JobWizardPageRerun;
