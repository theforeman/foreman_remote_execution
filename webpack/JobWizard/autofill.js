import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { get } from 'foremanReact/redux/API';
import { HOST_IDS, REX_FEATURE } from './JobWizardConstants';
import './JobWizard.scss';

export const useAutoFill = ({
  fills,
  setFills,
  setSelectedTargets,
  setHostsSearchQuery,
  setJobTemplateID,
  setTemplateValues,
  setAdvancedValues,
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (Object.keys(fills).length) {
      const {
        'host_ids[]': hostIds,
        search,
        feature,
        template_id: templateID,
        ...rest
      } = { ...fills };
      setFills({});
      if (hostIds) {
        const hostSearch = Array.isArray(hostIds)
          ? `id = ${hostIds.join(' or id = ')}`
          : `id = ${hostIds}`;
        dispatch(
          get({
            key: HOST_IDS,
            url: '/api/hosts',
            params: {
              search: hostSearch,
            },
            handleSuccess: ({ data }) => {
              setSelectedTargets(currentTargets => ({
                ...currentTargets,
                hosts: (data.results || []).map(({ id, name }) => ({
                  id,
                  name,
                })),
              }));
            },
          })
        );
      }
      if (search && !hostIds?.length) {
        setHostsSearchQuery(search);
      }
      if (templateID) {
        setJobTemplateID(+templateID);
      }
      if (feature) {
        dispatch(
          get({
            key: REX_FEATURE,
            url: `/api/remote_execution_features/${feature}`,
            handleSuccess: ({ data }) => {
              setJobTemplateID(data.job_template_id);
            },
          })
        );
      }
      if (rest) {
        Object.keys(rest).forEach(key => {
          const re = /inputs\[(?<input>.*)\]/g;
          const input = re.exec(key)?.groups?.input;
          if (input) {
            if (typeof rest[key] === 'string') {
              setTemplateValues(prev => ({ ...prev, [input]: rest[key] }));
              setAdvancedValues(prev => ({ ...prev, [input]: rest[key] }));
            } else {
              const { value, advanced } = rest[key];
              if (advanced) {
                setAdvancedValues(prev => ({
                  ...prev,
                  templateValues: { ...prev.templateValues, [input]: value },
                }));
              } else {
                setTemplateValues(prev => ({ ...prev, [input]: value }));
              }
            }
          }
        });
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fills]);
};
