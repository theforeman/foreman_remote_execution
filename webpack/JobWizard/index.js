import React from 'react';
import { Title, Divider } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import PageLayout from 'foremanReact/routes/common/PageLayout/PageLayout';
import { JobWizard } from './JobWizard';

const JobWizardPage = () => {
  const title = __('Run job');
  const breadcrumbOptions = {
    breadcrumbItems: [
      { caption: __('Jobs'), url: `/jobs` },
      { caption: title },
    ],
  };
  return (
    <PageLayout
      header={title}
      breadcrumbOptions={breadcrumbOptions}
      searchable={false}
    >
      <React.Fragment>
        <Title headingLevel="h2" size="2xl">
          {title}
        </Title>
        <Divider component="div" />
        <JobWizard />
      </React.Fragment>
    </PageLayout>
  );
};

export default JobWizardPage;
