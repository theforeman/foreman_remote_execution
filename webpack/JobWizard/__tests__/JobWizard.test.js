import React from 'react';
import * as patternfly from '@patternfly/react-core';
import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import JobWizardPage from '../index';
import { JobWizard } from '../JobWizard';

jest.spyOn(patternfly, 'Wizard');
patternfly.Wizard.mockImplementation(props => <div>{props}</div>);
const fixtures = {
  'renders ': {},
};
describe('JobWizardPage', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(JobWizardPage, fixtures));
});

describe('JobWizard', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(JobWizard, fixtures));
});
