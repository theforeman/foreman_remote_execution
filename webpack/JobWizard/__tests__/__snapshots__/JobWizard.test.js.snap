// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobWizard rendering renders  1`] = `
<mockConstructor
  className="job-wizard"
  height="70vh"
  navAriaLabel="Run Job steps"
  onClose={[Function]}
  steps={
    Array [
      Object {
        "component": <ConnectedCategoryAndTemplate
          category=""
          jobTemplate={null}
          setCategory={[Function]}
          setJobTemplate={[Function]}
        />,
        "name": "Category and template",
      },
      Object {
        "canJumpTo": false,
        "component": <p>
          TargetHosts 
        </p>,
        "name": "Target hosts",
      },
      Object {
        "canJumpTo": false,
        "component": <p>
           AdvancedFields 
        </p>,
        "name": "Advanced fields",
      },
      Object {
        "canJumpTo": false,
        "component": <p>
          Schedule
        </p>,
        "name": "Schedule",
      },
      Object {
        "canJumpTo": false,
        "component": <p>
          ReviewDetails
        </p>,
        "name": "Review details",
        "nextButtonText": "Run",
      },
    ]
  }
/>
`;

exports[`JobWizardPage rendering renders  1`] = `
<PageLayout
  breadcrumbOptions={
    Object {
      "breadcrumbItems": Array [
        Object {
          "caption": "Jobs",
          "url": "/jobs",
        },
        Object {
          "caption": "Run job",
        },
      ],
    }
  }
  header="Run job"
  searchable={false}
>
  <Title
    headingLevel="h2"
    size="2xl"
  >
    Run job
  </Title>
  <Divider
    component="div"
  />
  <JobWizard />
</PageLayout>
`;
